src/database.ts

```diff
@@ -55,24 +66,109 @@ import {
   TransactionRunner,
 } from './transaction-runner';
 
-export interface GetDatabaseOptions {
-  autoCreate?: boolean;
+import {google} from '../proto/spanner';
+import {Schema, RequestCallback, PagedRequest} from './common';
+import {ServiceError, CallOptions} from 'grpc';
+import {Readable, Transform, Duplex} from 'stream';
+import {PreciseDate} from '@google-cloud/precise-date';
+
+interface CreateBatchTransactionCallback {
+  (
+    err?: ServiceError | null,
+    transaction?: BatchTransaction | null,
+    apiResponse?: google.spanner.v1.ITransaction | google.spanner.v1.ISession
+  ): void;
 }
-export type DatabaseResponse = [Database, r.Response];
-export interface DatabaseCallback {
-  (err: Error | null, database?: Database, apiResponse?: r.Response): void;
+
+type CreateBatchTransactionResponse = [
+  BatchTransaction,
+  google.spanner.v1.ITransaction | google.spanner.v1.ISession
+];
+interface Cancellable {
+  cancel(): void;
 }
+type CancellableDuplex = Duplex & Cancellable;
+type DatabaseResponse = [Database, r.Response];
+interface DatabaseCallback {
+  (
+    err: ServiceError | null,
+    database?: Database,
+    apiResponse?: r.Response
+  ): void;
+}
+type GetSnapshotCallback = RequestCallback<Snapshot>;
+
+type GetTransactionCallback = RequestCallback<Transaction>;
 
-export interface GetSnapshotCallback {
-  (err: Error, snapshot?: null): void;
-  (err: null, snapshot: Snapshot): void;
+export interface SessionPoolConstructor {
+  new (
+    database: Database,
+    options?: SessionPoolOptions | null
+  ): SessionPoolInterface;
 }
 
-export interface GetTransactionCallback {
-  (err: Error, transaction?: null): void;
-  (err: null, transaction: Transaction): void;
+interface UpdateSchemaCallback {
+  (
+    err: ServiceError | null,
+    operation: GaxOperation,
+    resp: databaseAdmin.longrunning.IOperation
+  ): void;
 }
 
+type UpdateSchemaResponse = [
+  GaxOperation,
+  databaseAdmin.longrunning.IOperation
+];
+
+interface PoolRequestConfig {
+  reqOpts: {
+    // tslint:disable-next-line: no-any
+    [key: string]: any;
+    session?: string;
+  };
+  client?: string;
+  method?: string;
+  gaxOptions: CallOptions;
+}
+
+interface PoolRequestCallback {
+  (err: ServiceError | null, ...args: Array<{} | null>): void;
```

Comment:
I could be wrong, but will there ever actually be more than one param? Seems like maybe this should be a generic that accepts the response type.

Reply:
Created Generic Request response as below:
`request: <T, R = void>(config: RequestConfig,callback: RequestCallback<T, R>) => void;`